core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -83,48 +89,81 @@ func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query
 		return nil, errors.Wrap(err, "failed to DeleteExpired in Observation()")
 	}
 
-	queryRows := make([]*s4.Row, 0)
+	returnObservation := func(rows []*s4.Row) (types.Observation, error) {
+		promReportingPluginsObservationRowsCount.WithLabelValues(c.config.ProductName).Set(float64(len(rows)))
+		return MarshalRows(convertRows(rows))
+	}
+
 	unconfirmedRows, err := c.orm.GetUnconfirmedRows(c.config.MaxObservationEntries, pg.WithParentCtx(ctx))
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to GetUnconfirmedRows in Observation()")
 	}
 
-	if uint(len(unconfirmedRows)) < c.config.MaxObservationEntries {
-		versionRows, err := UnmarshalQuery(query)
+	if uint(len(unconfirmedRows)) >= c.config.MaxObservationEntries {
+		return returnObservation(unconfirmedRows[:c.config.MaxObservationEntries])
+	}
+
+	maxRemainingRows := int(c.config.MaxObservationEntries) - len(unconfirmedRows)
+	remainingRows := make([]*s4.Row, 0)
+
+	queryRows, err := UnmarshalQuery(query)
+	if err != nil {
+		c.logger.Errorw("Failed to unmarshal query (likely malformed)", "err", err)
+	} else {
+		snapshot, err := c.orm.GetSnapshot(s4.NewFullAddressRange(), pg.WithParentCtx(ctx))
```

Comment:
This needs to have the same address range as query.

Reply:
Agreed, this will require to change query structure, will change in the next iteration.